src/unity/toolkits/activity_classification/activity_classifier.cpp

```diff
@@ -726,46 +795,80 @@ gl_sframe activity_classifier::perform_inference(data_iterator *data) const {
-    // TODO: Implement double buffering.
-    std::map<std::string, shared_float_array> results =
-        backend->predict({ { "input", inputs.features } });
+    while (pending_batches.size() > remaining) {
 
-    // Copy the (float) outputs to our (double) buffer and add to the SArray.
-    const shared_float_array& output = results.at("output");
-    for (size_t i = 0; i < inputs.batch_info.size(); ++i) {
-      shared_float_array output_chunk = output[i];
-      data_iterator::batch::chunk_info info = inputs.batch_info[i];
+      // Pop one batch from the queue.
+      result batch = pending_batches.front();
+      pending_batches.pop();
+      
+      for (size_t i = 0; i < batch.data_info.batch_info.size(); ++i) {
+      
+      shared_float_array output_chunk = batch.output_info[i];
+      data_iterator::batch::chunk_info info = batch.data_info.batch_info[i];
 
       // Interpret the NN output as a sequence of NUM_PREDICTIONS_PER_CHUNK
       // probability vectors.
       ASSERT_EQ(output_chunk.size(), NUM_PREDICTIONS_PER_CHUNK * num_classes);
 
       const float* output_ptr = output_chunk.data();
       size_t cumulative_samples = 0;
-      while (cumulative_samples < info.num_samples) {
-        // Copy the probability vector for this prediction.
-        std::copy(output_ptr, output_ptr + num_classes, preds.begin());
-        output_ptr += num_classes;
-
-        // Compute how many samples this prediction applies to.
-        size_t num_samples = std::min(prediction_window,
-                                      info.num_samples - cumulative_samples);
-        cumulative_samples += prediction_window;
-        // Add a row to the output SFrame.
-        writer.write({ info.session_id, preds, num_samples },
-                     /* segment_id */ 0);
+      
+        while (cumulative_samples < info.num_samples) {
+          
+          // Copy the probability vector for this prediction.
+          std::copy(output_ptr, output_ptr + num_classes, preds.begin());
+        
+          output_ptr += num_classes;
+          
+          // Compute how many samples this prediction applies to.
+          size_t num_samples = std::min(prediction_window,
+                                        info.num_samples - cumulative_samples);
+          cumulative_samples += prediction_window ;
+          // Add a row to the output SFrame.
+          writer.write({ info.session_id, preds, num_samples },
+                       /* segment_id */ 0);
+          
+        }
       }
     }
+  };
+
+  while (data->has_next_batch()) {
+
+    data_iterator::batch input_batch =
+        data->next_batch(read_state<flex_int>("batch_size"));
+
+    // Wait until we have just one asynchronous batch outstanding. The work
+    // below should be concurrent with the neural net inference for that batch.
+    pop_until_size(1);
+    
+    result result_batch;
+
+    // Instead of giving the ground truth data to the image augmenter and the
```

Comment:
Let's clean up the comments. There is no image augmenter in the activity classifier case. Remember to take a look through any code/comments you copy and paste!